remove chimeMetadata

rhaas80 opened the pull request:
this pull request adds json backends for some metada items and generically removes use of chimeMetadata from all code.

It is not a full solution since it does not move all items into json.

rhaas80 committed:
jsonMetadata: JSON based science metadata

rhaas80 committed:
chordMetadata: add code to get frequncy indices for frame

rhaas80 committed:
BeamMetadata: use chordMetadata

rhaas80 committed:
chordMetadata: add nfreq, dataset_id, gps_time fields

rhaas80 committed:
BeamExtract: use chordMetadata

rhaas80 committed:
BeamInspect: use chordMetadata derived BeamMetadata

rhaas80 committed:
chordMetadata: add setter for dset_id

rhaas80 committed:
chordMetadata: add for for rfi_num_bad_inputs

rhaas80 committed:
RfiFrameDrop: use chordMetadata

rhaas80 committed:
jsonMetadata: add rfi_num_bad_frames

rhaas80 committed:
bufferBadInputs: use chordMetadata

rhaas80 committed:
HFBFrameView: add note about multiple frequencies per packet

rhaas80 committed:
chordMetadata: add support for lost timesamples

rhaas80 committed:
jsonMetadata: add lost_timesamples field

rhaas80 committed:
chordMetadata: make horrible hack to update json atomically compile

rhaas80 committed:
jsonMetadata: add timestamp elements

rhaas80 committed:
chordMetadata: add setters for fpga_seq_num, coarse_freq, gps_time

rhaas80 committed:
chordMetadata: add settter for lost_timesamples

rhaas80 committed:
chordMetadata: add timestamp of first package received

rhaas80 committed:
accumulate: use chordMetadata

rhaas80 committed:
beamformingPostProcess: use chordMetadata

rhaas80 committed:
compressLostSamples: use chordMetadata

rhaas80 committed:
frbPostProcess: use chordMetadata

rhaas80 committed:
compressLostSamples: fix function call

rhaas80 committed:
visBuffer: use chordMetadata

rhaas80 committed:
jsonMetadata: add rfi_flagged_samples

rhaas80 committed:
visTransform: use chordMetadata

rhaas80 committed:
chordMetadata: add rfi_flagged_samples

rhaas80 committed:
visAccumulate: use chordMetadata

rhaas80 committed:
testDataGen: use only chordMetadata

rhaas80 committed:
oneHotMetadata: remove chimeMetadata

rhaas80 committed:
nDiskFileWrite: use chordMetadata

rhaas80 committed:
basebandDumpData: use chordMetadata

rhaas80 committed:
DPDKShuffleSimulate: use chordMetadata

rhaas80 committed:
GenHFBFrames: use chordMetadata

rhaas80 committed:
networkOutputSim: use chordMetadata

rhaas80 committed:
pulsarSimProcess: use chordMetadata

rhaas80 committed:
testDataGenFloat: use chordMetadata

rhaas80 committed:
testDataGenQuad: use chordMetadata

rhaas80 committed:
fakeGpuPattern: use chordMetadata

rhaas80 committed:
fakeGpu: use chordMetadata

rhaas80 committed:
chordMetaData: add set_rfi_flagged_samples

rhaas80 committed:
chimeMetadataDump: use chordMetadata

rhaas80 committed:
metadataFactory: remove support for chimeMetadata

rhaas80 committed:
zeroSamples: use chordMetadata

rhaas80 committed:
iceBoardShuffle: remove unused chimeMetadata

rhaas80 committed:
chimeMetadataDump: remove stream_id

rhaas80 committed:
iceBoardStandard: use chordMetadata

rhaas80 committed:
invalidateVDIFframes: use chordMetadata

rhaas80 committed:
clBeamformKernel: use chordMetadata

rhaas80 committed:
fakeGpuPattern: use chordMetadata

rhaas80 committed:
pyPlotN2: remove use of chimeMetadata data

rhaas80 committed:
ICETelescope: remove get/set streamid functions

rhaas80 committed:
chimeMetadata: remove unused chimeMetadata files

rhaas80 committed:
testDataGen: restore stream_id

rhaas80 committed:
ReadGain: get freq_id from metadata rather than stream_id

rhaas80 committed:
frbPostProcess: get frequencies from metadata

rhaas80 committed:
Telescope: remove use of chimeMetadata

rhaas80 committed:
Telescope: remove buffer to freq_id mappings

instead clients should get them from metadata

rhaas80 committed:
clBeamformPhaseData: remove use of chimeMetadata

rhaas80 committed:
basebandReadout: get frequencies from metadata

rhaas80 committed:
RfiFrameDrop: get frequencies from metadata not buffer

rhaas80 committed:
jsonMetadata: add stream_id

rhaas80 committed:
chordMetadata: support stream_id

jbmertens committed:
update runner.py (#1275)

rhaas80 committed:
test out CI system

rhaas80 committed:
fakeTelescope: actually support time

rhaas80 committed:
fakeGpuPattern: set coarse frequency and lost samples

rhaas80 committed:
fakeGpuPattern: set frequencies

rhaas80 committed:
fakeGpuPattern: no longer set frequencies

andrerenard commented on lib/metadata/jsonMetadata.hpp:
```diff
@@ -0,0 +1,151 @@
+namespace jsonMetadata {
+// these must match chimeMetadata and choordMetadata for now
+// in namespace for C++ const
+#define MAX_NUM_BEAMS 20
```
This max beams logic was largely in place to deal with the fact we used structs which had compile time defined size.   We can remove that now that we have a dynamic structure.

rhaas80 replied:
the `max_num_beam` define is still used in the to-bytes serialized structure (`chordMetadataFmt` or so I think)

andrerenard commented on lib/metadata/jsonMetadata.hpp:
```diff
@@ -0,0 +1,151 @@
+// these must match chimeMetadata and choordMetadata for now
+// in namespace for C++ const
+#define MAX_NUM_BEAMS 20
+const int CHORD_META_MAX_FREQ = 1024;
```
Same with this parameter.

rhaas80 replied:
yup

andrerenard commented on lib/metadata/chordMetadata.hpp:
```diff
@@ -392,6 +399,128 @@ class chordMetadata : public metadataObject {
+        metadata[jsonMetadata::LOST_TIMESAMPLES] = lost_timesamples;
+    }
+
+    void atomic_add_lost_timesamples(const int32_t lost_samples) {
```
We need to rethink metadata changes where more than one producer could be setting values.   This is very much an edge case, but still needs to be accounted for.   My vote would be to add a mutex to the metadata object and then we can do this safely.

rhaas80 replied:
lost samples is currently one such where there are multiple threads accumulating onto the same metadata :-(

rhaas80 replied:
this one is the one that actually causes a SEGFAULT (due the cast to `atomic_int*` I assume).

rhaas80 replied:
added a `mutex` but this is still not thread safe unless `nlohmann_json` itself is thread safe.

andrerenard replied:
Hmm, that's a good question about `nlohmann_json`.  I don't think it's modification safe because in theory one of the underlying structures could change shape in the middle of a store.   Which actually creates a larger problem for all possible co-current update/insertions.  This means we need to be very strict about doing deep copies anytime we have pipeline divergence (although that was always the case, but for different reasons).  And multiple producer systems need to lock against all possible concurrent updates (or only have one thread set the metadata).      I think the updated the packet capture code to only set metadata from one thread, but I need to double check that...  

Having said all that, I think updating a value like this is safe with the lock, as long as the underlying structure doesn't change shape mid-store.   I'll need a bit more time to audit to code to see if this problem will actually show up anywhere...

The one issue I can see actually is that normally the value would be zeroed in metadata structure, so setting an initial 0 value wasn't needed.   However here the first call could throw an exception for a missing value unless it's initialized to zero in the code that calls it.   We could fix that either by updating the calling code, or checking for the value's existence before setting, and initializing to zero if needed.

rhaas80 replied:
I added protections using a `mutex` (standard one so far, so not recursive) to all getters and setters. Waiting for the CI tests to finish now.

andrerenard commented on lib/metadata/chordMetadata.hpp:
```diff
@@ -392,6 +399,128 @@ class chordMetadata : public metadataObject {
+    }
+
+private:
+    jsonMetadata::metadata metadata;
```
We will need serialization/deserialization functions for this class.   Specifically for stages like `bufferSend`/`bufferRecv` we currently only perform "shallow copies" which worked for the flat struct model, but don't work with this dynamic structure.   Of course for CHIME compatibility we'll also want to have a "legacy" version, which will likely have some conflicts with what Jim is working on now...

rhaas80 replied:
true, well it now inherits from `MetadataObject` so there's a serialize function to overload

rhaas80 replied:
I have kept the serialize to bytes functionality intact (and working) though, so if the buffers actually used that (they should...) then they'd still work.

andrerenard commented on lib/metadata/chordMetadata.hpp:
```diff
@@ -392,6 +399,128 @@ class chordMetadata : public metadataObject {
         this->onehot_index[dim] = i;
         strncpy(this->onehot_name[dim], name.c_str(), CHORD_META_MAX_DIMNAME);
     }
+
```
General comment on this class.   We now seem to have a slightly odd mix of old struct based logic for some items, and new json backed methods.  This is going to have some strange effects with serialization/deserialization.   As much as it's a huge pain, my feeling is we need to either commit entirely to changing this all over, or at the very least create serialization/deserialization which addresses _both_ the old components and the new components. 

Also in terms of naming.  I think we should rename this `frameMetadata`, since it's basically just taking on all possible types of frame metadata in the system.   We might someday want to breakup the classes just for clarity, but that's not obviously an improvement.

rhaas80 replied:
yup, for anything to go forward this needs to all move into json. This pull request (and the work for it) was mostly to see if there are already objections in how the getters / setters work. I will move the leftover bits to json.

rhaas80 replied:
the onehot code was removed, with approval by @dstndstn

andrerenard commented on lib/metadata/chordMetadata.hpp:
```diff
@@ -392,6 +399,128 @@ class chordMetadata : public metadataObject {
+    /// should NOT be used, use @c lost_timesamples instead.
+    /// This value will be filled even if RFI zeroing is disabled.
+    int32_t get_rfi_flagged_samples() const {
+        return metadata[jsonMetadata::RFI_FLAGGED_SAMPLES].template get<int32_t>();
```
What does the runtime error message look like when a get value doesn't exist?  Should we create a generic `get()` function that creates nice error messages about where things failed, and then throws the exception in case the caller wants to handle the case of no value set.

rhaas80 replied:
I'd expect it would be a bus-error (if eg the atomic `int` needs alignment that a regular `int` violates). Really what we want is a mutex, such as the one removed in cd1d4ebf .  :-)

rhaas80 replied:
I think it's a C++ exception, out of range is what the docs say: https://json.nlohmann.me/home/exceptions/#jsonexceptionout_of_range403

andrerenard commented on lib/metadata/jsonMetadata.hpp:
```diff
@@ -0,0 +1,151 @@
+        coarseFreq(const metadata& md) {
+        md[NFREQ].get_to(nfreq);
+        const auto& freq(md[COARSE_FREQ]);
+        if(freq.size() > MAX_NUM_BEAMS)
```
Shouldn't this be max frequencies not beams.   But also I think we should be able to get rid of hard coded max sizes since we are backing the metadata with a dynamic structure.

andrerenard reviewed (commented):
Not finished the review, but leaving some early comments.  

I think we are going to have to take a careful look at places where metadata is generically handled.   Specifically `bufferSend/bufferRecv`, some of the file writing logic, and the buffer functions `copy_metadata` and `move_metadata` which used to rely on being able to shallow copy the structure.   The current test environment is mostly geared to testing individual stages, which I think is why we aren't seeing these changes throw errors.  That should change with some of the work @jbmertens is doing with the CI self-hosted runners, but I don't think has very good coverage of these generic parts of the framework at the moment.    It might be worth having some simple metadata passing pipelines get added to the `config/ci-tests` if that's now working.  

We also need to consider how to address all the metadata types used by "CHORD" (and thus the kernels we need for CHIME).  To merge this in, we are going to need to either create some hybrid serialization logic, or commit to fully converting all the "CHROD" metadata types to use json under the hood, and then serialize that.

rhaas80 committed:
iceBoard: remove chimeMetadata

rhaas80 committed:
N2Accumulate: remove direct use of chordMetadata fields

rhaas80 committed:
chordMetadata: remove fields that have migrated to json already

rhaas80 committed:
chordMetadata: use from_json for beamCoord

rhaas80 committed:
jsonMetadata: use from_json for coarseFreq

rhaas80 committed:
chordMetadata: use direct json access

rhaas80 committed:
gpuSimulateN2k: use chordMetadata

rhaas80 committed:
chordMetadata: use json for coarse_freq

rhaas80 committed:
jsonMetadata: remove redundant nfreq entry

rhaas80 committed:
accululate: adapt to changes in chordMetaData::coarse_freq

rhaas80 committed:
BeamExtract: adapt to chordMetaData::coarse_freq change

rhaas80 committed:
BeamMetaData: adapt to chordMetadata::coarse_freq change

rhaas80 committed:
jsonMetadata: remove unused coarse_freq struct

rhaas80 committed:
chimefrb_template: adjust to chordMetadata::coarse_freq

rhaas80 committed:
frb_template: adjust to chordMetadata::coarse_freq

rhaas80 committed:
cudaRFIS012: adjust to chordMetadata::nfreq

rhaas80 committed:
cudaCorrelator: adjust to chordMetadata::nfreq

rhaas80 committed:
N2Metadata: adjust to chordMetadata::nfreq

rhaas80 committed:
fakeGpu: adjust to chordMetadata::coarse_freqs

rhaas80 committed:
asdfFileRead: adjust to chordMetadata::coarse_freq

rhaas80 committed:
asdfFileWrite: adjust to chordMetadata::coarse_freq

rhaas80 committed:
gdalFileRead: adjust to chordMetadata::coarse_freq

rhaas80 committed:
gdalFileWrite: adjust to chordMetadata::coarse_freq

rhaas80 committed:
hdf5FileRead: adjust to chordMetadata::coarse_freq

rhaas80 committed:
hdf5FileWrite: adjust to chordMetadata::coarse_freq

rhaas80 committed:
FEngine: adjust to chordMetadata::coarse_freq

rhaas80 committed:
BeamInspect: adjust to chordMetadata::coarse_freq

rhaas80 committed:
testDataGen: adjust to chordMetadata::coarse_freq

rhaas80 committed:
cudaPL1bitCorrelator: adjust to chordMetadata::coarse_freq

rhaas80 committed:
cudaPLMaskExpander: adjust to chordMetadata::coarse_freq

rhaas80 committed:
cudaRFISKtilde: adjust to chordMetadata::coarse_freq

rhaas80 committed:
gpuSimulateN2kPLExpand: adjust to chordMetadata::coarse_freq

rhaas80 committed:
gpuSimulateN2kCorr: adjust to chordMetadata::coarse_freq

rhaas80 committed:
gpuSimulateN2kPL1bitCorr: adjust to chordMetadata::coarse_freq

rhaas80 committed:
cudaPL1bitCorrelator: adjust to chordMetadata::coarse_freq

rhaas80 committed:
chordMetadata: add jsonMetadata member

rhaas80 committed:
pulsarPostProcess: start adding chordMetadata

rhaas80 committed:
jsonMetadata: introduce sample0_offset and offset_downsampling

rhaas80 committed:
chordMetadata: use jsonMetadata for smaple0_offset and offset_downsampling

rhaas80 committed:
cudaBasebandBeamformer_chime: adjust to smaple0_offset change

rhaas80 committed:
cudaBasebandBeamformer_chord: adjust to smaple0_offset change

rhaas80 committed:
cudaBasebandBeamformer_hirax: adjust to smaple0_offset change

rhaas80 committed:
cudaBasebandBeamformer_pathfinder: adjust to smaple0_offset change

rhaas80 committed:
cudaCopyFromRingbuffer: adjust to smaple0_offset change

rhaas80 committed:
cudaCopyToRingbuffer: adjust to smaple0_offset change

rhaas80 committed:
cudaCorrelator: adjust to smaple0_offset change

rhaas80 committed:
cudaFRBBeamReformer: adjust to smaple0_offset change

rhaas80 committed:
cudaPL1bitCorrelator: adjust to smaple0_offset change

rhaas80 committed:
asdfFileRead: adjust to smaple0_offset change

rhaas80 committed:
asdfFileWrite: adjust to smaple0_offset change

rhaas80 committed:
hdf5FileRead: adjust to smaple0_offset change

rhaas80 committed:
hdf5FileWrite: adjust to smaple0_offset change

rhaas80 committed:
testDataGen: adjust to smaple0_offset change

rhaas80 committed:
gpuSimulateN2kCorr: adjust to smaple0_offset change

rhaas80 committed:
gpuSimulateN2kPL1bitCorr: adjust to smaple0_offset change

rhaas80 committed:
gpuSimulateN2kPLExpand: adjust to smaple0_offset change

rhaas80 committed:
chordMetadata: remove unused lost_fpga_samples, rfi_flagged_samples

rhaas80 committed:
gdalFileWrite: adjust to smaple0_offset change

rhaas80 committed:
gdalFileRead: adjust to smaple0_offset change

rhaas80 committed:
N2Accumulate: remove us of lost_fpga_samples[f][vis_samp_n]

use of lost_fpga_samples[f][vis_samp_n] removed while Geoff udpates
n2Accumulate

rhaas80 committed:
FEngine: adjust to smaple0_offset change

rhaas80 committed:
chordMetadata: remove oneHost elements

rhaas80 committed:
testDataGen: remove chordMetadata::one_hot code

rhaas80 committed:
chordMetadata: correct check for coarseFreq size

rhaas80 committed:
chordMetadata: move half_fpga_sample0 etc into json metadata

rhaas80 committed:
jsonMetadata: add freq_upchan_factor etc

rhaas80 committed:
frb_template: adjust to chordMetadata::freq_upchan_factor

rhaas80 committed:
bb: adjust to chordMetadata::sample0_offset

rhaas80 committed:
chimefrb_template: adjust to chordMetadata::freq_upchan_factor

rhaas80 committed:
cudaPLMaskExpander: adjust to chordMetadata::freq_upchan_factor

rhaas80 committed:
cudaRFIS012: adjust to chordMetadata::freq_upchan_factor

rhaas80 committed:
cudaRFIS012bar: adjust to chordMetadata::freq_upchan_factor

rhaas80 committed:
cudaPL1bitCorrelator: adjust to chordMetadata::time_downsampling_fpga

rhaas80 committed:
cudaCorrelator: adjust to chordMetadata::time_downsampling_fpga

rhaas80 committed:
asdFileWrite: adjust to chordMetadata::half_fpga_sample0 etc

rhaas80 committed:
asdFileRead: adjust to chordMetadata::half_fpga_sample0 etc

rhaas80 committed:
cudaRFISKtilde: adjust to chordMetadata::half_fpga_sample0

rhaas80 committed:
gpuSimulateN2kPL1bitCorr: freq_upchan_factor

rhaas80 committed:
gpuSimulateN2kCorr: adjust to chordMetadata::time_downsampling_fpga

rhaas80 committed:
gpuSimulateN2kPLExpand: adjust to chordMetadata::freq_upchan_factor

rhaas80 committed:
hdf5FileRead: adjust to chordBase::freq_upchan_factor etc

rhaas80 committed:
hdf5FileWrite: adjust to chordMetadata::freq_upchan_factor etc

rhaas80 committed:
testDataGen: adjust to chordMetadata freq_upchang_factor etc

rhaas80 committed:
gdalFileWrite: adjust to chordMetadata::freq_upchan_factor

rhaas80 committed:
gdalFileRead: adjust to chordMetadata::freq_upchan_factor

rhaas80 committed:
FEngine: adjust to chordMetadata::freq_upchan_factor

rhaas80 committed:
jsonMetadata: add frame_counter

rhaas80 committed:
chordMetadata: move frame_counter to json

rhaas80 committed:
FEngine: adjust to chordMetadata::frame_counter

rhaas80 committed:
iceBoardShuffle: re-add missing metadata

got removed in initial committ for upgraded hardware

rhaas80 committed:
cudaCopyNtoRingbuffer: use chordMetadata::coarse_freq

rhaas80 commented:
@andrerenard so tests are failing on the CI system due to CI system issues it seems.

Since I have re-formatted and auto-generated code in here, should I create two pull requests:

1. to actually be reviewed without auto-gen and reformat, this will always fail the tests
2. to actually commit, incl. the auto-gen and reformatting, this is about 400 changed files and lots of noise

rhaas80 committed:
chordMetadata: re-add mutex for atomic update of lost_timesamples

rhaas80 committed:
chordMetadata: add placebo lock for lost_timesamples count

this is not actually thread save unless nlohmann_json is

andrerenard commented on lib/metadata/BeamMetadata.hpp:
```diff
@@ -29,8 +30,8 @@ class BeamMetadata : public metadataObject {
-    /// Stream identifier
-    stream_t stream_id;
+    /// Frequency indices in frame
+    std::vector<int> coarse_freq;
```
We might want to leave this as a `stream_t` actually.  We only really use this metadata in the `beamExtract` stage, which will get updated to make it's output identical to what it outputs in CHIME now.   As a result we may need to reverse lookup the appropriate streamIDs.   Since we have not yet made our way to this part of the project, we could just leave this as `stream_t` and return to generating a valid stream_id in BeamExtract later.  With a big TODO note in that code.

rhaas80 replied:
ok. I'll give it a try once the review is done. Not sure if I actually have a `stream_id` to copy in given that data that is available to `BeamMetadata`.

andrerenard commented on lib/core/metadataFactory.cpp:
```diff
@@ -84,11 +83,6 @@ std::shared_ptr<metadataPool> metadataFactory::new_pool(const std::string& pool_
                                     pool_type);
     }
 
-    if (pool_type == "chimeMetadata") {
```
Other than to make sure things compiled without CHIME metadata, I'm not sure there is value to removing it entirely.  We are going to use this structure eventually in the interface layers.   Perhaps rename to `legacyChimeMetadata`?

rhaas80 replied:
likely, yes. There will not be any factory produced chimeMetadata though. Only the struct that is produced for legacy buffer writers (if there's any that need to write chimeMetadata). In which case I will revert the commit that deletes the files I guess.

andrerenard commented on lib/metadata/chordMetadata.hpp:
```diff
@@ -80,59 +89,11 @@ class chordMetadata : public metadataObject {
-    // FPGA samples.
-    int time_downsampling_fpga[CHORD_META_MAX_FREQ];
-
     // Dish layout
     int ndishes;                                  // number of dishes
     int n_dish_locations_ew, n_dish_locations_ns; // number of possible dish locations
```
In the interest of this change not being too crazy large, I'm ok with leaving these kind of variables for now, but we should really factor this out to the telescope object.

andrerenard commented on lib/metadata/chordMetadata.hpp:
```diff
@@ -50,28 +56,31 @@ class chordMetadata : public metadataObject {
-    /// Track the number of rfi-flagged samples in each gpu sub-integration
-    int rfi_flagged_samples[CHORD_META_MAX_FREQ][CHORD_META_MAX_VIS_SAMPLES];
-
     int dims;
     int dim[CHORD_META_MAX_DIM];
     char dim_name[CHORD_META_MAX_DIM][CHORD_META_MAX_DIMNAME]; // "F", "T", "D", etc
```
TODO note, I think this should move into the buffer class in some capacity so we don't need to carry it around like this, but that is another project...

rhaas80 replied:
using the non-data stuff of NDArray preferably I'd say.

andrerenard reviewed (changes requested):
I'm a bit hung-up on the atomic issue + wider possible JSON thread safety issues.  My current read is that it's actually ok given the way things are currently used.  However I need to spend a few more hours double checking the code.  It's likely the case that we need to add some software protection against metadata misuse to catch ourselves if we do something unsafe...  Possibly by limiting how stages are allowed to mutate metadata, ideally without just a big mutex on everything, but I need to think about this a bit more.

rhaas80 commented:
hmm, a copy `metadata_a = metadata_b` would work, but a `memcpy` which is likely what it does will not. Hmm it copies using `metadataObject` shared pointers. I *am* suspicious:

```
void GenericBuffer::private_copy_metadata(int dest_frame_id, GenericBuffer* src, int src_frame_id) {
    std::shared_ptr<metadataObject> from_metadata_container = src->metadata[src_frame_id];
    std::shared_ptr<metadataObject> to_metadata_container = metadata[dest_frame_id];
    if (from_metadata_container->get_object_size() != to_metadata_container->get_object_size()) {
        WARN("Metadata sizes don't match, cannot copy metadata!!");
        return;
    }
    *to_metadata_container = *from_metadata_container;
}
```

jbmertens committed:
networkOutputSim: remove var per clang warning

jbmertens committed:
testDataGen: add frame counter

rhaas80 committed:
iceBoardShuffle: initialize lost samples and rfi counters

jbmertens committed:
cudaPLMaskExpander: fix propagating downsampling factors

rhaas80 committed:
chordMetadata: protect all accesses with a mutex

rhaas80 commented:
superseded by #1308
